### ISSUE TITLE
rfc: Strongly Typed Authorization Area and Sessions

### ISSUE THREAD

**Issue by oddcoder** [tpm-rs/tpm-rs#87]
Latest RFC Iteration may be found [here](https://github.com/tpm-rs/tpm-rs/issues/87#issuecomment-2308938715)

## Summary

This issue investigates strongly typed implementation of the *authorization
area* and *sessions* for TPM commands.

## Motivation

The Rust type system is robust enough to build a correct-by-construction
*authorization area* with minimal boilerplate code. This makes it syntactically
impossible to:

1. Create an authorization area with more than three sessions.
2. Pass an authorization area to commands that do not expect it.

Additionally, the Rust type system allows:

3. Specializing an *authorization area* type that requires a minimum of one or
   two *sessions* for commands that require them. This specialization prevents
   the passing of incomplete authorization areas when using this specific type.
4. Simplifying the task of validating an *authorization area*.

### Session

We define the general session type via a trait as follows:

```rust
pub trait Session {
    fn get_auth_command(&self) -> TpmsAuthCommand;
    fn validate_auth_response(&self, auth: &TpmsAuthResponse) -> TpmResult<()>;
}
```

We also define a specific trait for sessions suitable for encryption and decryption. The trait's form is not yet clear, It will most like not require `alloc` based vector. But the idea will be along the lines of:

```rust
pub trait EncDec: Session {
	fn decrypt (cipher_text: &[u8]) -> Vec<u8>;
	fn encrypt (cipher_text: &[u8]) -> Vec<u8>;

}
```

### Authorization Area
We establish the infrastructure for the *authorization area*. Within this framework, we define four distinct types for the *authorization area*, corresponding to zero, one, two, and three *sessions*, respectively
```rust
pub struct AuthorizatinArea0;
pub struct AuthorizationArea1<T: Session>(T);
pub struct AuthorizationArea2<T: Session, U: Session>(T, U);
pub struct AuthorizationArea3<T, U, V>(T, U, V)
    where T: Session,
          U: Session,
          V: Session;
```

*Session* types are resolved statically at compile time. This eliminates the runtime overhead typically associated with dynamic dispatch. If dynamic dispatch is necessary, we can explicitly implement the `Session` trait for types like `Box<dyn Session>` or other similar variants.

Due to the static nature of `AuthorizationAreaX`, we cannot use a single type `T: Session` for all sessions. For example, in the case of `AuthorizationArea3`, it is possible that all three sessions have completely different types. All *authorization areas* have their own constructor.

Sometimes it is desired to have lower bound on many  *sessions* are available for an arbitrary *authorization area* type `A` for that we define traits for for lower bound. In particular we define the following traits:

```rust

/// Authorization area with no information on how many sessions it has
pub trait AA<T, U, V> {
    fn decompose(self) -> (Option<T>, Option<U>, Option<V>);
}
/// Authorization area with 1+ sessions
pub trait AA1P<T, U, V> {
    fn decompose(self) -> (T, Option<U>, Option<V>);
}
/// Authorization area with 2+ sessions
pub trait AA2P<T, U, V> {
    fn decompose(self) -> (T, U, Option<V>);
}
```

To simplify things there is auto trait implementations in particular there is:

```rust
impl <T, U, V, W> AA1P<T, U, V> for W where W: AA2P<T, U, V>;
impl <T, U, V, W> AA<T, U, V> for W where W: AA1P<T, U, V>;
```

That drastically reduces the number of traits we need to implement per each `AuthorizatinAreaX`

The user of `client` will always construct concrete types such as 
`AuthorizationArea1`, `AuthorizationArea2`, `AuthorizationArea3`
While in `client` and `services` no concrete types would be used and it will always be `impl AA`, `impl AA1P`, `impl AA2P`.

## Drawbacks

The strongly-typed approach to *authorization areas* and *sessions* may be complex and not the best choice for a reference implementation. This is especially true for anyone trying to create a custom TPM implementation in a language with a weaker type system than Rust. They would have to figure out all the invariants and guarantees encoded in the type structure of *sessions*.

It is also worth noting that the implementation might require a step back and some redesign if TCG decide to extent the *authorization area* section to contain more than three sessions. 

## Alternatives

1- Instead of custom types for Authorization area, we can just use tuples.
2- we can define `AA`, `AA1P`, `AA2P` as concrete types instead of traits.

Both ideas sound better.

## Unresolved Questions

- Is `AuthorizatinArea0`, and `AA` necessary ?
- Should *authorization areas* own the sessions
- Is it desirable to define `AuthorizationArea2`, `AuthorizationArea3` in terms of `AuthorizationArea1`, `AuthorizationArea2` ?
- How will this infrastructure integrate with the new RPC client?
- Is the overhead in the text section within acceptable margins for the requirements set by TPM chip vendors?
- Can we enforce the mandatory use of a session, ensuring that neglecting a session triggers a compile-time error?

### tpm-rs

The current rust implementation defines the interface for interacting with each session via the trait `Session`:

```rust
pub trait Session {
    fn get_auth_command(&self) -> TpmsAuthCommand;
    fn validate_auth_response(&self, auth: &TpmsAuthResponse) -> TpmResult<()>;
}
```

The current implementation represents *authorization area* as a vector with fixed capacity:

```rust
pub type CmdSessions<'a> = ArrayVec<&'a mut dyn Session, 3>;
```

This vector is not dynamically allocated. Attempting to pass more than three sessions will result at runtime error. This is handled implicitly by `ArrayVec`.

### go-tpm

Like `tpm-rs`, `go-tpm` defines sessions through an interface, but it offers a more comprehensive implementation.

```go
type Session interface {
	Init(tpm transport.TPM) error
	CleanupFailure(tpm transport.TPM) error
	NonceTPM() TPM2BNonce
	NewNonceCaller() error
	Authorize(cc TPMCC, parms, addNonces []byte, names []TPM2BName, authIndex int) (*TPMSAuthCommand, error)
	Validate(rc TPMRC, cc TPMCC, parms []byte, names []TPM2BName, authIndex int, auth *TPMSAuthResponse) error
	IsEncryption() bool
	IsDecryption() bool
	Encrypt(parameter []byte) error
	Decrypt(parameter []byte) error
	Handle() TPMHandle
}
```

The implementation distinguishes between sessions suitable for encryption, decryption, and those that are suitable for neither. This distinction is made dynamically at runtime.

Whenever possible, The *authorization area* is primarily defined as a list of variadic arguments of the type `Session`. If this is not possible, a slice of the type `Session` is used instead. This approach provides minimal implicit guarantees at both compile time and runtime. As a result, it is the developer's responsibility to ensure that the correct number of sessions is always provided.

### tpm2-tss

The `tpm2-tss` manages each session as a linked list of metadata:

```c
typedef struct RSRC_NODE_T {
    ESYS_TR esys_handle;
    TPM2B_AUTH auth;
    IESYS_RESOURCE rsrc;
    size_t reference_count;
    struct RSRC_NODE_T * next;
} RSRC_NODE_T;
```

It is not clear what is the point of maintaining a linked list. Especially that *authorization area* is defined as

```c
RSRC_NODE_T *session_tab[3];
```

All checks are explicit, written in the implementation.

### ms-tpm-20-ref

The representation of a *session* here is the most compelete. It is

```c
typedef struct SESSION
{
    SESSION_ATTRIBUTES attributes;
    UINT32 pcrCounter;
    UINT64 startTime;
    UINT64 timeout;
    CLOCK_NONCE epoch;
    TPM_CC commandCode;
    TPM_ALG_ID authHashAlg;
    TPMA_LOCALITY commandLocality;
    TPMT_SYM_DEF symmetric;
    TPM2B_AUTH sessionKey;
    TPM2B_NONCE nonceTPM;
    union
    {
        TPM2B_NAME boundEntity;
        TPM2B_DIGEST cpHash;
        TPM2B_DIGEST nameHash;
        TPM2B_DIGEST templateHash;
    } u1;
    union
    {
        TPM2B_DIGEST auditDigest;
        TPM2B_DIGEST policyDigest;
    } u2;
} SESSION;
```

It is important to note that this representation of a *session* includes information for both client communication and the execution of TPM operations on the TPM implementation itself.

## POC
https://github.com/tpm-rs/tpm-rs/pull/95

**Comment by trent-reed** [comment]
Sorry for the delay here, I'll take a look at this RFC early next week - this is probably going to take a bit of thought for me. Thanks for putting this together! :)

**Comment by oddcoder** [comment]
little spoiler alert! I am doing small modifications to the design as I go along with prototyping!

**Comment by trent-reed** [comment]
Boy, sessions are complicated. I'm currently still sifting through the specification and forming my responses. I'm very sorry about the delay here, this is currently my top-priority and I am actively working on feedback, it's just that sessions seem to be a very complicated part of the spec.

I'll ask for just a few more days, and if it starts to take too long I'll just throw caution to the wind and give early feedback.

**Comment by oddcoder** [comment]
## Summary

This issue investigates strongly typed implementation of the _authorization
area_ and _sessions_ for TPM commands.

## Motivation

The Rust type system is robust enough to build a correct-by-construction
_authorization area_ with minimal boilerplate code. This makes it syntactically
impossible to:

1. Create an authorization area with more than three sessions.
2. Pass an authorization area to commands that do not expect it.

Additionally, the Rust type system allows:

3. Specializing an _authorization area_ type that requires a minimum of one or
   two _sessions_ for commands that require them. This specialization prevents
   the passing of incomplete authorization areas when using this specific type.
4. Simplifying the task of validating an _authorization area_.

## Problem with [Last Design](https://github.com/tpm-rs/tpm-rs/issues/87#issue-2458009390)

1- Authorization areas are defined as custom opaque types which might be unnecessary.
2- `AuthorizatinArea0` is un-necesssary
3- `AA`, `AA1P`, `AA2P` are defined as traits, and they can be just defined as concrete types.
4- The way traits were implemented, required the user to write annotations about the expected
types of the non-existing sessions.

## Explanation

### Session

We define the general session type via a trait as follows:

```rust
pub trait Session {
    fn get_auth_command(&self) -> TpmsAuthCommand;
    fn validate_auth_response(&self, auth: &TpmsAuthResponse) -> TpmResult<()>;
}
```

We also define a specific trait for sessions suitable for encryption and decryption. The trait's form is not yet clear, It will most like not require `alloc` based vector. But the idea will be along the lines of:

```rust
pub trait EncDec: Session {
	fn decrypt (cipher_text: &[u8]) -> Vec<u8>;
	fn encrypt (cipher_text: &[u8]) -> Vec<u8>;

}
```

### Authorization Area

We establish the infrastructure for the _authorization area_. Within this framework, we use:
1- Unit type `()` for empty authorization areas.
2- Single session for _authorization area_ with one session.
3- Tuples for the _authorization area_ corresponding two and three _sessions_.

_Session_ types are resolved statically at compile time. This eliminates the runtime overhead typically associated with dynamic dispatch. If dynamic dispatch is necessary, we can explicitly implement the `Session` trait for types like `Box<dyn Session>` or other similar variants.

Sometimes it is desired to have lower bound on many _sessions_ are available for an arbitrary _authorization area_ type `A` for that we define traits for for lower bound. In particular we define the following types for _authorization area_ with lower bounds:

```rust
/// Authorization area with no information on how many sessions it has
pub struct AA<T, U, V> (Option<T>, Option<U>, Option<V>);
/// Authorization area with 1+ sessions
pub struct AA1P<T, U, V> (T, Option<U>, Option<V>);
/// Authorization area with 2+ sessions
pub struct AA2P<T, U, V> (T, U, V);
```

To simplify matters, we will have trait implementations that allow certain conversion, this will allow us to talk about certain
:

```rust
// conversion to AA
impl From<()> for AA<PasswordSession, PasswordSession, PasswordSession>;
impl <T> From<T> for AA<T, PasswordSession, PasswordSession> where T: Session;
impl <T, U> From<(T,U)> for AA<T, U, PasswordSession>;
impl <T, U, V> From<(T,U, V)> for AA<T, U, V>
// conversion to AA1P
impl From<T> for AA1P<T, PasswordSession, PasswordSession> where T: Session;
impl<T, U> From<(T,U)> for AA1P<T, U, PasswordSession>;
impl<T, U, V> From<(T,U, V)> for AA1P<T, U, V>;
// conversion to AA2P
impl<T, U> From<(T,U)> for AA2P<T, U, PasswordSession>;
impl<T, U, V> From<(T,U, V)> for AA2P<T, U, V>;
```

Additionally we define way conversion between `AA`, `AA1P` and `AA2P` as follows:

```rust
// conversions that lose information about minimum numbers of sessions
impl<T, U, V> From<AA1P::<T,U, V>> for AA<T, U, V>;
impl<T, U, V> From<AA2P::<T,U, V>> for AA<T, U, V>;
impl<T, U, V> From<AA2P::<T,U, V>> for AA1P<T, U, V>;
```

## Client Example
The following snippet shows an example of a function that expects
arbitrary number of sessions (up to three):

```rust
pub fn run_command_with_handles<CmdT, T, X, Y, Z>
(
    cmd: &CmdT,
    cmd_handles: CmdT::Handles,
    cmd_sessions: impl Into<AA::<X, Y, Z>>,
    tpm: &mut T,
) -> TpmResult<(CmdT::RespT, CmdT::RespHandles)>
where
    CmdT: TpmCommand,
    T: Tpm,
    X: Session,
    Y: Session,
    Z: Session;
```

## Unresolved Questions
- Should _authorization areas_ own the sessions?
- How will this infrastructure integrate with the new RPC client?
- Ho would we use this sessions implementation in services?
- Is the overhead in the text section within acceptable margins for the requirements set by TPM chip vendors?
- Can we enforce the mandatory use of a session, ensuring that neglecting a session triggers a compile-time error?

**Comment by trent-reed** [comment]
Sorry for the long delay here, I had to really dig into sessions to understand them better. Here's what I came up with: https://gist.github.com/trent-reed/2fd9f8e4a9a0dc387ed493915b5e0beb

From this understanding, I'll leave some feedback on this RFC in my next comment.

**Comment by trent-reed** [comment]
* I'm not sure there's practical justification in separating the encryption/decryption from the session type. All inputs we need to provide may be usable for parameter encryption/decryption, so I'm not seeing what separating the two provides for us, or if it's even possible.
  * It's really more of a special case that password sessions cannot be used for enc/dec/audit.
* The current abstraction proposed doesn't really allow us to specify how one intends to use the session. A session could be usable for audit, but maybe we don't want to audit (similar for encryption/decryption). So I think we're missing something there in saying "how I want to use this session".
  * Auth is obviously figurable from the command being run and the index of the session in the session list, but the other parameters - not so much.
* We're missing the `name` component, which is a hidden dependency to produce the command authorization (see Part 1, 18.7: Command Parameter Hash). The client needs to be able to provide this along with the handle for this hash. 
  * Note: The server also needs this, but it can produce it from the handle provided, as it knows the name of the loaded object. It's more like a check to see that the client is interacting with the expected handle.
* There's actually some mutable state that's important in a session (`nonceTpm`, and `nonceCaller`), I believe the current design might not account for keeping sessions past a single command since it takes by value. We might have to do some `SessionRef<'a>` shenanigans or something.
* I think it might be nice to couple the authorized sessions with the authorized handles (kinda like go-tpm does), so that they're listed right next to one-another.
* My main issue is I don't find these types to be very readable for what they do, and the more I read about the additional sessions and the session limits of 3 the more I realize it's probably best to not be too perscriptive on that limit of 3 here...
  * I think a limit of 3 additional sessions is reasonable if it's easy to code that somewhere, but personally I think the best thing to do is:
    * If the command only allows `TPM_ST_NO_SESSIONS`, no sessions are allowed, optional or otherwise.
    * If the command allows session we require `N` where `N` is the number of authorized handles.
    * We can always accept <=3 additional handles regardless of this number `N`, because that limit is only stated in the spec because of minimum session requirements for a TPM implementation.
    * Why 3? Because outside of authorization, the only other things you can use sessions for are: encryption, decryption, and audit - all of these are allowed to be separate sessions.
* We probably need to work on our session type abstraction too. There's some state we need to manage, and probably wise to wrap up calling the `TPM2_StartAuthSession` in these helper types.
  * My ideal state here is that we provide helper types to call these TPM commands for you where needed, but then if one wants to create their own session types they can still call into the underlying commands to accomplish this. This is kinda how go-tpm is setup, and I think I like it.

---

I'm actually kind of wondering, this might be another case where a proc-macro is nicest. You could imagine something really nice looking like this (and imagine this new `Command` trait has functions like `create_authorization_section` and `create_parameter_section` or whatever - basically helpers the client can call):

```rs
#[derive(Command)]
#[command(code=..., tag=TPM_ST_SESSIONS, response=CreateResponse)]
struct CreateCommand {
  #[authorization(index=1, role=user)]
  pub parent_handle: AuthorizedHandle,
  pub in_sensitive: Tpm2bSensitiveCreate,
  pub in_public: Tpm2bPublic,
  pub outside_info: Tpm2bData,
  pub creation_pcr: TpmlPcrSelection,
}

#[derive(Marshalable)]
struct CreateResponse {
  pub out_private: Tpm2bPrivate,
  pub out_public: Tpm2bPublic,
  pub creation_data: Tpm2bCreationData,
  pub creation_hash: Tpm2bDigest,
  pub creation_ticket: TpmtTkCreation,
}

// Then elsewhere...

let command = CreateCommand {
  parent_handle: AuthorizedHandle {
    handle: target_parent_handle,
    name: ...,
    authorization: target_authorization,
  },
  other_params: {
    ...
  },
};
let response = client.run_command(&command, additional_sessions)?;
```

**Comment by trent-reed** [comment]
Oh, also IDK if it's worth our first attempt at this cutting out dynamic dispatch... Let's just get something working first, then we'll see if we can make it static.

Dynamic dispatch is possible without `Box` for clients who can't allocate things with `&dyn Trait`, so it shouldn't be impossible.

**Comment by oddcoder** [comment]
> I'm not sure there's practical justification in separating the encryption/decryption from the session type.

Neither, honestly, I am not focusing on that now, the thing to note is that if we want to separate, it is clear that we can do both. But least for this part, I will try to focus only the one problem which is strongly tying for session usage.

>  We might not be able to get all these checks at complie-time, that's fine - let's just do what we can. If making a check compile-time comes at the cost of making it harder to read or use, then we ought to have a discussion about this in a case-by-case basis.

I think the current **latest** design is extremely simple and covers all cases, with the exception of linking sessions and handles, I will have to check how go-tpm handles it and will propose something in the next iteration.

**Comment by oddcoder** [comment]
> The current abstraction proposed doesn't really allow us to specify how one intends to use the session

I am struggling to see how can this be enforced at compile time. and I cannot see the value. If you have an idea for the use-case please let me know.

**Comment by trent-reed** [comment]
> I am struggling to see how can this be enforced at compile time. and I cannot see the value.

It can't be; I just don't see how the proposed abstraction provides a way of controlling this, and we're definitely going to run into this.

I guess we could have a separation here, where there's a type that maintains the session, and that can be different from the type that implies to the client how to use a session (and what session to use). So, perhaps there's a path forward with your proposed code, but I think we'll probably need something like that.

---

About the proposed design. My major feedback is basically:

* I really don't understand what you're trying to say with the "to AA" and "to AA1P" conversions. That looks like a codesmell to me, but I'd have to see an example using it to know what I really think.
* I don't like `TpmCommand::Handles` in the original abstraction, which this design is trying to re-use... This puts no limitation on the types that can be claimed to be provided, and relies on the command implementer to fill it in with only the known-good types (handles). (Basically I'm claiming we can do better to enforce the proper types here.)
* I don't see where the client provides the `name` information for computing the Command Parameter Hash (maybe the separate "how to use the session" types could help with this, so maybe I'd demote this to the minor feedback below).
* I wish the authorization sessions could be closer to the handles that they authorize (kind of like how the go implementation does it - that implementation isn't perfect with what it does for sessions, but I do like that aspect of it).
* I'm not sure how the proposed abstraction will work for commands that explicitly don't allow sessions (even optional ones).
  * To be clear I think this will almost always be some kind of runtime check, but I'm not seeing where we encode that in the command structure in the proposed solution, is my point.

Some more minor feedback, though I'm less concerned about this because I do see a path forward for it:

* Not really a fan of the type names here. Names like `AA` and `AA1P` sound like they'll be very confusing when I see them in code.
* We probably have to consider the other use-cases of sessions when constructing this abstraction, otherwise we're bound to re-write it when we go to implement those features (e.g. encryption, decryption, audit).
* Probably we'll need some way of telling the client whether or not to persist a session (I think your abstraction allows for this via a proxy reference type, though we may need to add some additional trait functions to suggest the intent to the client code).

---

All that said, none of it's blocking. I don't want to hold this up on "this is not how I'd implement this", so while I'm just not the biggest fan of the design, I'll give the LGTM on it.

I do think it might be nice to have one example of a simple command like `Create` to show how all this links together.

---

I'd rather we did something more structured and easy to read like what I mentioned in my comment above.

```rs
#[derive(Command)]
#[command(code=TpmCc::Create, supports_sessions=true, response=CreateResponse)]
struct CreateCommand {
  #[authorization(index=1, role=user)]
  pub parent_handle: AuthorizedHandle,
  pub in_sensitive: Tpm2bSensitiveCreate,
  pub in_public: Tpm2bPublic,
  pub outside_info: Tpm2bData,
  pub creation_pcr: TpmlPcrSelection,
}
```

This:

* doesn't bake in a limit of the number of sessions in the code, we enforce it at proc-macro time, which will be easy to change.
  * Frankly, I'm leaning more and more towards just not enforcing any limit of 3 at compile time, and instead leaving the compile-time checking to be solely around what *must* be provided (e.g. this command has 2 authorized handles, thus it must have at least 2 sessions). Then we'd leave the TPM to respond to the user based on whether or not they used too many sessions. I suspect the spec is overzealous in suggesting the limit of 3 is important for the client.
* allows us to enforce invariants about the types we expect through the proc-macro (e.g. handles must always be placed before parameters, and they must be this specific type)
* check the input where possible (like index numbers)
* react to commands that must support no sessions explicitly (we bake that reaction in from the data provided to the proc-macro)
* provides a really clear readable type, self-documenting structure for handle authorization.
* removes the boilerplate of implementing `TpmCommand` everywhere.

It doesn't exactly describe the fine details of some of the things I mentioned above (like how session persistence works, for example), but like with your design I also see a path forward here, so those are maybe minor nits to iron out as we implement it.

---

TL;DR: I'm not the biggest fan of the proposal, but I don't want to block on "perfect", so LGTM.

**Comment by oddcoder** [comment]
Hello @trent-reed . I tried to work a bit doing a third revision of the sessions but I didn't like that it became too complicated, it had the potential of solving problems, but those problems are not visible now. 

I want to point out that there are currently two design, [here](https://github.com/tpm-rs/tpm-rs/issues/87#issue-2458009390) and [here](https://github.com/tpm-rs/tpm-rs/issues/87#issuecomment-2308938715)

I think you comments are regarding the first, the second should be much much simpler. The purpose of is just to solve one problem that is easy to test which is avoiding less than necessary arguments.

I would welcome more modifications, but simple the current tests would not make it possible to try. If things feel bad but I didn't create them (working with Handle/Command traits, mutability of sessions) I try not to touch them until I have good reason to do. I will finalize the implementation based the second proposal all while picking sensible naming docs and example in the draft PR.

However, if you would like to discuss the design, or implementation or anything else in online meeting to facilitate communication, also please let me know.

**Comment by trent-reed** [comment]
Hey @oddcoder, my comments were in relation to the second (newer) design.

Anyways, as I mentioned, we don't need to block on perfect. We can always come back and take a second look at these things in the future. So I'm fine with moving forward with this for now, but there are aspects that I think this current setup doesn't handle, and I think some other designs for this could be cleaner and harder to misuse.

But yeah, doesn't need to be perfect first-go. So LGTM for now, and we'll improve things as we go.
